### ISSUE TITLE
Parser.oneOf could use expanded documentation of when order matters

### ISSUE THREAD

**Issue by morgen-peschke** [typelevel/cats-parse#184]
It appears that order matters for `Parser.oneOf`, in cases where one parser accepts a subset of another parser. 

This makes sense, however it might be good to explicitly mention in the docs the implication this has for generating parsers for a set of `String` values - specifically that they should be reverse sorted according to length, because it's very easy to create inconsistent parsers if the input isn't correctly prepared.

For example, parsing a truthy value for `true` will consistently work (or fail) depending on the order of the parsers:
```scala
import cats.parse.{Parser => P}

val buggy: P[Boolean] =
  P.oneOf(List("1", "t", "tru", "yes", "true").map(P.string(_)))
    .void
    .as[Boolean](true)

val works: P[Boolean] =
  P.oneOf(List("true", "yes", "tru", "t", "1").map(P.string(_)))
    .void
    .as[Boolean](true)

def sort(strings: List[String]): List[String] =
  strings
    .map(str => (str.length, str))
    .sorted
    .reverse
    .map(_._2)

val sorted: P[Boolean] =
  P.oneOf(sort(List("1", "t", "tru", "yes", "true")).map(P.string(_)))
    .void
    .as[Boolean](true)

List("1", "t", "tru", "true", "y")
  .foreach { input =>
    println {
      """|%-6s => %6s => %s
         |%-6s => %6s => %s
         |%-6s => %6s => %s
         |""".stripMargin.format(
           s"<$input>", "buggy", buggy.parseAll(input),
           "", "works", works.parseAll(input),
           "", "sorted", sorted.parseAll(input)
         )
    }
  }
```

This is particularly troublesome because the error looks like an unexpected end of string, rather than an expected end of string that didn't happen:
```
Left(Error(1,NonEmptyList(EndOfString(1,3))))
```
[Scastie](https://scastie.scala-lang.org/OP8AVkBlROGYyvue66wM5w)

**Comment by johnynek** [comment]
I'm actually don't see the error here.

oneOf should always care about ordering. It tries to match the left first, then moves to the right.

There are optimizations to improve performance, which could have bugs, but the design is to to match the left first then the right. So, if you match `t` or `true` the true will never match because `t` has already matched.

Can you write a unit test that you think should pass that does not?

The reason you see end of string, if you try to match `true` against `tru`, then you will hit the end of the string and have an error with that.

**Comment by morgen-peschke** [comment]
I agree the library is not the source of the error, this is more of a request for a documentation improvement. 

The correct way to sort a list of strings you're going to be parsing isn't obvious if you're not looking at them, so it's an easy way to introduce a subtle bug in your parser if you're not building it with a hard-coded list.

This is a particularly easy trap to fall into because, if you're porting a parser from `fastparse` (or just used to it's semantics), it's easy to miss that `P.oneOf(List(...).map(P.string(_))` isn't equivalent to `StringIn(...)`, [which would work](https://scastie.scala-lang.org/3yYGOglTRRa8p9BBEGVvCQ).

The context I ran into this was I needed to parse a value that matched a protobuf enum with about a hundred entries, and the naive approach (`P.oneOf(Companion.values.map(_.name).map(P.stringIgnoreCase(_))`) worked about 80% of the time.

Probably something for the wiki or the Migrating from Fastparse section, I'm not really sure what to suggest as the documentation appears to be in flux at the moment.
